client/core/bond.go

```diff
@@ -0,0 +1,768 @@
+				} else {
+					liveBonds = append(liveBonds, bond)
+				}
+			}
+			return liveBonds
+		}
+
+		dc.acct.authMtx.Lock()
+		// Screen the unexpired bonds slices.
+		dc.acct.bonds = filterExpiredBonds(dc.acct.bonds)
+		dc.acct.pendingBonds = filterExpiredBonds(dc.acct.pendingBonds) // possible expired before confirmed
+		// Extract the expired bonds.
+		expiredBonds := make([]*db.Bond, len(dc.acct.expiredBonds))
+		copy(expiredBonds, dc.acct.expiredBonds)
+		dc.acct.authMtx.Unlock()
+
+		if len(expiredBonds) == 0 {
+			continue
+		}
+
+		spentBonds := make([]*bondID, 0, len(expiredBonds))
+		for _, bond := range expiredBonds {
+			bondIDStr := fmt.Sprintf("%v (%s)", coinIDString(bond.AssetID, bond.CoinID), unbip(bond.AssetID))
+			if now < int64(bond.LockTime) {
+				c.log.Debugf("Expired bond %v refundable in about %v.",
+					bondIDStr, time.Duration(int64(bond.LockTime)-now)*time.Second)
+				continue
+			}
+
+			assetID := bond.AssetID
+			wallet, err := c.connectedWallet(assetID)
+			if err != nil {
+				c.log.Errorf("%v wallet %v not available to refund bond %v: %v",
+					unbip(bond.AssetID), bondIDStr, err)
+				continue
+			}
+			if _, ok := wallet.Wallet.(asset.Bonder); !ok { // will fail in RefundBond, but assert here anyway
+				c.log.Errorf("Wallet %v is not an asset.Bonder", unbip(bond.AssetID))
+				return
+			}
+
+			expired, err := wallet.LockTimeExpired(ctx, time.Unix(int64(bond.LockTime), 0))
+			if err != nil {
+				c.log.Errorf("Unable to check if bond %v has expired: %v", bondIDStr, err)
+				continue
+			}
+			if !expired {
+				c.log.Debugf("Expired bond %v with lock time %v not yet refundable according to wallet.",
+					bondIDStr, time.Unix(int64(bond.LockTime), 0))
+				continue
+			}
+
+			// Generate a refund tx paying to an address from the currently
+			// connected wallet, using bond.PrivKey to create the signed
+			// transaction. The RefundTx is really a backup.
+			priv := secp256k1.PrivKeyFromBytes(bond.PrivKey)
+			newRefundTx, err := wallet.RefundBond(ctx, bond.Version, bond.CoinID, bond.Data, bond.Amount, priv)
+			bondAlreadySpent := errors.Is(err, asset.CoinNotFoundError) // or never mined!
+			if err != nil && !bondAlreadySpent {
+				c.log.Errorf("Failed to generate bond refund tx: %v", err)
+				continue
+			}
+
+			// If the user hasn't already manually refunded the bond, broadcast
+			// the refund txn. Mark it refunded and stop tracking regardless.
+			if bondAlreadySpent {
+				c.log.Warnf("Bond output not found, possibly already spent or never mined! "+
+					"Marking refunded. Backup refund transaction: %x", bond.RefundTx)
+			} else {
+				refundCoinID, err := wallet.SendTransaction(newRefundTx)
+				if err != nil {
+					c.log.Errorf("Failed to broadcast bond refund txn %v: %v", newRefundTx, err)
+					continue
+				}
+				c.log.Infof("Bond %v refunded in %v (%s)", bondIDStr, coinIDString(assetID, refundCoinID), unbip(bond.AssetID))
```

Comment:
`[INF] CORE: Bond ca9a6c08dc4c973a0524b32a05fc9bf9b28182bed39223508aef190b18d229e3:0 (dcr) refunded in 0c21facb520a6e9ae79e81dbd9860a9c7e958b951cb12947e93a96b4c2ef07a3:0 (dcr)`